### PR TITLE
port block cache and compression

### PR DESCRIPTION
**Pull request by bobotu**
cherry-pick block cache and block compression

### CODE REVIEW

File: table/table.go [hunk 1]
```diff
@@ -27,16 +27,28 @@ import (
...
+	"github.com/dgraph-io/ristretto"
+	"github.com/golang/snappy"
 	"github.com/pingcap/errors"
 )
 
 const fileSuffix = ".sst"
+const intSize = int(unsafe.Sizeof(int(0)))
+
+// TableInterface is useful for testing.
+type TableInterface interface {
```
Comment (coocood) [inline table/table.go]: Not used?

File: table/table.go [hunk 2]
```diff
@@ -266,29 +283,34 @@ func (t *Table) readIndex() {
 	t.blockEndOffsets = bytesToU32Slice(buf)
 }
 
-func (t *Table) block(idx int) (block, error) {
-	y.Assert(idx >= 0)
+func (t *Table) block(idx int) (*block, error) {
+	y.AssertTruef(idx >= 0, "idx=%d", idx)
```
Comment (coocood) [inline table/table.go]: AssertTruef is slower.

File: table/table.go [hunk 3]
```diff
@@ -266,29 +283,34 @@ func (t *Table) readIndex() {
 	t.blockEndOffsets = bytesToU32Slice(buf)
 }
 
-func (t *Table) block(idx int) (block, error) {
-	y.Assert(idx >= 0)
+func (t *Table) block(idx int) (*block, error) {
```
Comment (coocood) [inline table/table.go]: We don't need to change to pointer.

File: table/builder.go [hunk 4]
```diff
@@ -204,28 +207,38 @@ func (b *Builder) addHelper(key []byte, v y.ValueStruct) {
...
+		data, err = b.compressData(b.buf)
+		y.Check(err)
+	}
+
+	if err := b.w.Append(data); err != nil {
+		return err
+	}
+	b.blockEndOffsets = append(b.blockEndOffsets, uint32(b.writtenLen+len(data)))
+	b.writtenLen += len(data)
+	b.blockBaseOffset = uint32(b.writtenLen + len(data))
```
Comment (coocood) [inline table/builder.go]: This value never used.

**Comment by bobotu** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**